### PR TITLE
[ZEPPELIN-6053] Fix(zeppelin-web): Upgrade webpack-dev-server to enable HMR

### PR DESCRIPTION
**Pull request by Henry-Hong**
### What is this PR for?
- Fix : https://issues.apache.org/jira/browse/ZEPPELIN-6052

Background
---
Currently, the Zeppelin web application uses Webpack for its build tools. Webpack provides valuable features like hot module replacement (HMR) through webpack-dev-server, which greatly enhances the development experience by automatically updating the UI in response to code changes.

Unfortunately, hot module replacement is not functioning properly, impacting the developer experience, particularly on the frontend (zeppelin-web).

Resolution
---

To resolve this issue, I updated webpack-dev-server from version ^3.1.10 to ^4.15.2. This update was selected as the latest version compatible with Webpack version 4.

- Version Upgrade: Although I considered further upgrades, I chose the latest version compatible with Webpack version 4. I plan to migrate to Webpack version 5 incrementally.
- Dependencies: The updated version 4.15.2 requires @webpack-cli/serve version ^1.2.0, which necessitated the installation of webpack-cli@4.x.x. Consequently, I installed webpack-cli@4.10.0 too.
- Configuration Changes: Due to BREAKING CHANGES introduced by the new version, modifications were required in the webpack.config.js to accommodate these updates. I followed [this migration guide](https://github.com/webpack/webpack-dev-server/blob/master/migration-v4.md).

AS-IS
---
화면 기록 2024-08-11 오후 3 14 14

TO-BE
---
화면 기록 2024-08-11 오후 1 57 45

### What type of PR is it?
Bug Fix

### Todos
* [x] Upgrade webpack-dev-server (3.1.10 -> 4.15.2)
* [x] Install @webpack-cli/serve 
* [x] Upgrade webpack-cli (3.3.12 -> 4.10.0)

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-6053

### How should this be tested?
```
1. git clone latest code.
git clone https://github.com/apache/zeppelin.git

2. change directory to zeppelin-web
cd zeppelin-web

3. install dependencies
npm install

4. run dev mode
npm run dev

5. modify any codes you want to see the change (I'd recommend <H1/> simply)
- 1. move to zeppelin-web/src/app/home.html
- 2. change any html codes you can easily detect

6. you can check it doesn't update UI automatically. you should reload instead.
```

### Questions:
* Does the license files need to update?
* Is there breaking changes for older versions?
* Does this needs documentation?

### CODE REVIEW

File: zeppelin-web/webpack.config.js [hunk 1]
```diff
@@ -274,24 +274,32 @@ module.exports = function makeWebpackConfig () {
...
+      devServer.app.use('**/app/', express.static(path.resolve(__dirname, './src/app/')));
+      devServer.app.use('**/assets/', express.static(path.resolve(__dirname, './src/assets/')));
+      devServer.app.use('**/fonts/', express.static(path.resolve(__dirname, './src/fonts/')));
+      devServer.app.use('**/components/', express.static(path.resolve(__dirname, './src/components/')));
+    },
+    devMiddleware: {
+      stats: 'minimal', // were moved to devMiddleware
     },
-    stats: 'minimal',
+    static: './src'
```
Comment (Reamer) [inline zeppelin-web/webpack.config.js:302]: Maybe
```suggestion
    static: {
      directory: './src'
    }
```

File: zeppelin-web/package.json [hunk 2]
```diff
@@ -40,50 +38,53 @@
...
+    "headroom.js": "^0.9.3",
     "highlight.js": "9.4.0",
     "jquery": ">=1.8",
     "jquery-ui": ">=1.8",
     "jquery.scrollto": "~2.1.2",
-    "diff": "3.3.0",
     "json3": "~3.3.1",
     "jszip": "2.6.1",
     "lodash": "~3.9.3",
+    "mathjax": "^2.7.9",
```
Comment (seung-00) [inline zeppelin-web/package.json:66]: It looks like the dependency version has changed. It seems to be related to #4786, so wouldn’t it be better to keep the existing version in this PR?

File: zeppelin-web/webpack.config.js (hunk 1 above)
Comment (Henry-Hong) [reply zeppelin-web/webpack.config.js:302]: b71cbf1ddc22f630dedf941c17820e5a7e8dfc74

**Comment by jongyoul** [comment]
@Henry-Hong Could you please resolve the conflict first?

**Comment by Henry-Hong** [comment]
> @Henry-Hong Could you please resolve the conflict first?

@jongyoul Conflict resolved! Thanks :-)

File: zeppelin-web/package.json
Comment (jongyoul) [inline zeppelin-web/package.json:1]: I like this kind of sorting :-) Thank you.

File: zeppelin-web/package.json [hunk 3]
```diff
@@ -40,50 +38,53 @@
...
+    "headroom.js": "^0.9.3",
     "highlight.js": "9.4.0",
     "jquery": ">=1.8",
     "jquery-ui": ">=1.8",
     "jquery.scrollto": "~2.1.2",
-    "diff": "3.3.0",
     "json3": "~3.3.1",
     "jszip": "2.6.1",
     "lodash": "~3.9.3",
+    "mathjax": "3.0.0",
```
Comment (Reamer) [inline zeppelin-web/package.json:66]: I thought mathjax should not be updated to 3.0.0. See https://github.com/apache/zeppelin/pull/4786

File: zeppelin-web/package.json [hunk 4]
```diff
@@ -40,50 +38,53 @@
     "angular-sanitize": "1.5.7",
     "angular-touch": "1.5.7",
     "angular-ui-ace": "0.2.3",
-    "angular-websocket": "1.0.14",
+    "angular-ui-bootstrap": "~2.5.0",
+    "angular-ui-grid": "4.4.6",
+    "angular-websocket": "1.1.0",
```
Comment (Reamer) [inline zeppelin-web/package.json:43]: Is this an intentional minor update?

Comment (Henry-Hong) [reply zeppelin-web/package.json]: No, it wasn't. 

I think there was a mistake during rebase. 

Fixed! 
https://github.com/Henry-Hong/zeppelin/blob/e2b12d1a5a53f17c65a87ac9997b294aeff946b0/zeppelin-web/package.json#L43

File: zeppelin-web/package.json (hunk 3 above)
Comment (Henry-Hong) [reply zeppelin-web/package.json]: You're right.

Check the followings.
I think there was a mistake during rebase.

https://github.com/Henry-Hong/zeppelin/blob/e2b12d1a5a53f17c65a87ac9997b294aeff946b0/zeppelin-web/package.json#L66

**Comment by Reamer** [comment]
If no further comments are received, I will merge this to the master on Friday.

Of course, the CI pipeline must be 'green'.

**Pull request merged**